pkg/agent/openflow/pipeline.go

```diff
@@ -1724,76 +1602,39 @@ func (c *client) snatFlows(nodeIP net.IP, localSubnet net.IPNet, category cookie
 		// covered by other flows (the flows matching the local and
 		// remote Pod subnets) anyway.
 		l3FwdTable.BuildFlow(priorityNormal).
-			MatchProtocol(binding.ProtocolIP).
+			MatchProtocol(ipProto).
 			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
 			MatchCTMark(gatewayCTMark, nil).
 			Action().GotoTable(nextTable).
 			Cookie(c.cookieAllocator.Request(category).Raw()).
 			Done(),
 
-		// Add the SNAT mark on the packet that is not filtered by other
-		// flow entries in the L3Forwarding table.
+		// Send the traffic to external to snatTable.
 		l3FwdTable.BuildFlow(priorityLow).
-			MatchProtocol(binding.ProtocolIP).
-			MatchCTStateNew(true).MatchCTStateTrk(true).
+			MatchProtocol(ipProto).
 			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
-			Action().LoadRegRange(int(marksReg), snatRequiredMark, snatMarkRange).
-			Action().GotoTable(nextTable).
+			Action().GotoTable(snatTable).
 			Cookie(c.cookieAllocator.Request(category).Raw()).
 			Done(),
-		// Force IP packet into the conntrack zone with SNAT. If the connection is SNATed, the reply packet should use
-		// Pod IP as the destination, and then is forwarded to conntrackStateTable.
-		c.pipeline[conntrackTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolIP).
-			Action().CT(false, conntrackStateTable, CtZone).NAT().CTDone().
+		// For the traffic tunneled from remote Nodes, rewrite the
+		// destination MAC to the gateway interface MAC.
+		l3FwdTable.BuildFlow(priorityLow).
+			MatchProtocol(ipProto).
+			MatchRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
```

Comment:
That is true. Do you think I should change the comment to list other possibilities?

Reply:
It would be great to have to help understand. I noticed this because the comment says it's for "the traffic tunneled from remote Nodes" but the code matches `macRewriteMark` instead of `markTrafficFromTunnel` which could have been symmetrical with the flow above. But now I wonder if locally generated traffic will be processed by this flow. It will randomly match the above flow and this one, right? Should we use `markTrafficFromTunnel` here to be deterministic.

Reply:
You are right. The uses of macRewriteMark become hard to track now! But in the case it is set by LB flows, do you see we should still rewrite dst MAC here? I am not familiar with LB flows to make a judgement.
Probably we should add more comments on LB flows about why macRewriteMark is set.

Reply:
I think LB flows need to set `macRewriteMark` because the original dest MAC is the gateway's MAC, after DNATed it should be updated to the Pod's MAC if the selected endpoint is a local Pod. It doesn't need to rewrite MAC in other cases. I think we should still rewrite dst MAC and match `markTrafficFromTunnel` here. Then local traffic will always hit the above flow and tunnel traffic will hit this one.

Reply:
Changed to match markTrafficFromTunnel.